### ISSUE TITLE
Models deforming when converting from Blender to GLTF

### ISSUE THREAD

**Issue by cgordon5025** [KhronosGroup/glTF-Blender-IO#2104]
**Describe the bug**

**To Reproduce**
1. Go to File>Export>GLTF
2. Click on ExportToGLTF
3. See error

**Expected behavior**
When exporting my scene with animated models, I expect that the models retain their structure and integrity as I see it in the blender GUI.

**Screenshots**
**Expected Behavior (screenshot taken in Blender):**
![Screenshot 2024-01-08 182344](https://github.com/KhronosGroup/glTF-Blender-IO/assets/112432755/0ec2326a-012a-4a36-acc4-328362c5219b)

**Actual Behavior (screenshot taken in React Project):**
![Screenshot 2024-01-08 182316](https://github.com/KhronosGroup/glTF-Blender-IO/assets/112432755/0baa31a6-3ce6-4833-914c-4732e0705d3a)

[myProject.zip](https://github.com/KhronosGroup/glTF-Blender-IO/files/13875387/myProject.zip)

**Version**
 - OS: Windows 11 Build 22621.2861
 - Blender Version: 3.4.1 Date 2022-12-19

**Additional context**
I am using the polmandres GLTF to JSX converter to render this, when using their provided GLTF viewer (https://gltf.pmnd.rs/) the models looks correct, yet in the project they are warped. Further when I export within Blender, I see immediately that the Fox and Dog models have the same deformation that appear in the gltf file. If I click undo, they are 'fixed', but the gltf remains warped.

**Comment by donmccurdy** [comment]
To help us narrow this down, do you see the same issue when running gltfjsx with a higher precision? The default is 3 decimals precision in node position/rotation/scale, which could be an issue. There's an option for this in the gltfjsx GUI, or the `--precision` flag in the CLI. I think the CLI is generally more up-to-date than the web GUI.

> Further when I export within Blender, I see immediately that the Fox and Dog models have the same deformation that appear in the gltf file...

I'm not sure I understood this part, do you mean you see the deformation _in Blender_ after export, unrelated to gltfjsx? I'm testing in Blender 4.0 and not seeing the deformation there so far.

**Comment by julienduroure** [comment]
Seems your rigs are using BBone, that can not be exported in glTF.
This can explain deformation issue in any glTF viewer.

Note that Blender now have the requested API to convert BBone to multiple bones. But the work still to be done to manage it correctly at export is huge, and not planned for now.

**Comment by cgordon5025** [comment]
> I'm not sure I understood this part, do you mean you see the deformation _in Blender_ after export, unrelated to gltfjsx? I'm testing in Blender 4.0 and not seeing the deformation there so far.

Thought I had resolved it by updating my blender to 4.0, things were working yesterday and now this morning it broke once again. Yes, both the gltf file and Blender have deformed bones after the export occurs. This morning I had to make a quick adjustment to the scene, and when I exported it once more the bones deformed again

**Comment by cgordon5025** [comment]
> Seems your rigs are using BBone, that can not be exported in glTF. This can explain deformation issue in any glTF viewer.
> 
> Note that Blender now have the requested API to convert BBone to multiple bones. But the work still to be done to manage it correctly at export is huge, and not planned for now.

I used rigify to create the rig. I am fairly new to Blender. Is there any way I can remove/change this without restarting the rig? Or a way to convert it myself before exporting?

**Comment by cgordon5025** [comment]
@julienduroure Another question, if it is the bendy bones why are the two other rigged models in this file work/retain their structure, while the two quadrupeds lose it?

**Comment by julienduroure** [comment]
If you want me to investigate more deeply, can you please create a new .blend file with only the dog and **a single action** for it, where we can see the deformation the best, and indicate what part is deformed.

As far as I can see, there is not so much BBone deformation on owl and crab (even if there is some BBones).

But there are for the dog (on the neck for example):

![image](https://github.com/KhronosGroup/glTF-Blender-IO/assets/1696020/52b653c7-ad03-4cd4-88d4-f007b9c85102)

**Comment by cgordon5025** [comment]
Sure I can provide that. What is odd, is that I made a demo version of this file with only one animal, and it exported without problem. I have since updated to Blender 4.0, but the issue is still persisting. I am noticing it is only the front paws of both the fox and dog (quadrupeds) that are affected, would changing the bone type for that segment resolve this, though in an unorthodox manner?
[DogCustomRig.zip](https://github.com/KhronosGroup/glTF-Blender-IO/files/13893926/DogCustomRig.zip)

**Comment by cgordon5025** [comment]
@julienduroure my Blender file is changed after export AND the deformation is viewed in the final destination. When using a GLTF viewer (https://gltf.pmnd.rs/) the issue does not show itself.

It seems to be a bone rotation issue, that after executing the export front_foot_heel)ik.R/L changes its orientation. Because the rig is parent to the model, the model deforms due to the front_foot_heel_ik.R/L changing orientation/rotation. 

CustomDogRig was showing no issues when I tried to export it. The issue is only in the main file (myProject) 

Below are photos of the model in three different environments
In blender (before exporting)
![inBlender](https://github.com/KhronosGroup/glTF-Blender-IO/assets/112432755/586f235a-956d-475f-b44f-ed71e5ffa758)


In GLTF Viewer

![inGLTFViewer](https://github.com/KhronosGroup/glTF-Blender-IO/assets/112432755/0f32cf6d-e92d-485c-9f2e-46bbcb3368e4)

In ReactProject and Blender after exporting
![inReactProject](https://github.com/KhronosGroup/glTF-Blender-IO/assets/112432755/62beca75-65bb-41c7-bf4c-56b5d484740d)

**Comment by julienduroure** [comment]
Seems I am not able to reproduce


Before Export :

Pose
![image](https://github.com/KhronosGroup/glTF-Blender-IO/assets/1696020/45669512-141f-48ee-9fe0-2050d40959f7)

Rest
![image](https://github.com/KhronosGroup/glTF-Blender-IO/assets/1696020/82fd9f10-9d88-4d59-9ab0-cb7111b47e02)


After Export:

![image](https://github.com/KhronosGroup/glTF-Blender-IO/assets/1696020/9b226a05-406d-4001-b935-7c3d0fc7ed07)

Pose
![image](https://github.com/KhronosGroup/glTF-Blender-IO/assets/1696020/6923a822-4a74-4fcb-9fa1-32b2a1b09136)

Rest
![image](https://github.com/KhronosGroup/glTF-Blender-IO/assets/1696020/3abc713c-14ad-41d3-874e-e0e6424cf5f6)


In https://gltf.pmnd.rs/ 

![image](https://github.com/KhronosGroup/glTF-Blender-IO/assets/1696020/a1a4a877-5354-4461-ac15-1a2439df327e)


In https://gltf-viewer.donmccurdy.com/

![image](https://github.com/KhronosGroup/glTF-Blender-IO/assets/1696020/b82c0194-b4eb-4a74-b48b-b24d70212226)

**Comment by cgordon5025** [comment]
@julienduroure This is correct, however noce you convert this to jsx and using gltfjsx, you will see that the animations are all deformed. the strangest part is that the neutral poses are correct, but once you apply the animation it does not work any more. 

Attached is the sandbox I have been using to test my animations in a React project

[animation-sandbox.zip](https://github.com/KhronosGroup/glTF-Blender-IO/files/13978814/animation-sandbox.zip)

**Comment by donmccurdy** [comment]
@cgordon5025 is it possible to observe the deformation in any of these viewers?

- https://gltf-viewer.donmccurdy.com/
- https://sandbox.babylonjs.com/
- https://gltf.report/

If not, I think it may be necessary to report an issue at https://github.com/pmndrs/gltfjsx/. This exporter aims to produce a valid and visually-correct glTF file, but generating a JSX template from that is an additional process with its own complexities.

**Comment by cgordon5025** [comment]
@donmccurdy Yes, in the first link provided, https://gltf-viewer.donmccurdy.com/

You can see that the dog, when in idle/rest position that the legs are proper and not deformed, but when activating the animation, the rig becomes deformed 

See attached video.

https://github.com/KhronosGroup/glTF-Blender-IO/assets/112432755/1b05ce65-858a-4856-9fd7-da8e14586d45

**Comment by julienduroure** [comment]
The legs are also with this shape in Blender before export , when using this action DogHeadNod

![image](https://github.com/KhronosGroup/glTF-Blender-IO/assets/1696020/89b23c26-5d42-4193-98cd-bc181a5afa9e)

**Comment by cgordon5025** [comment]
In my blender file they are not like that, let me send another
[playground.zip](https://github.com/KhronosGroup/glTF-Blender-IO/files/13979029/playground.zip)


I have noticed that if it is not deformed before exporting, it will become deformed, but then you are able to undo w/ control-z

**Comment by julienduroure** [comment]
I can reproduce with this new .blend file
Investigation in progress

**Comment by julienduroure** [comment]
I opened a discussion here about the Blender API to reset bones to rest pose (that is root of the issue):

https://projects.blender.org/blender/blender/issues/118077

**Comment by julienduroure** [comment]
Some bones are locked and have non default values (MCH-front_foot_parent.L for example). During switch between actions, to avoid unexpected issue when some channels are keyed on an action, but not in another, we reset bone matrices. This reset make all matrices Identity. But because of non default locked bone, this is not the behavior we get when using Clear Transforms (alt+G/R/S).

You can workaround that (if all your channels are keyed) by disable this option: Reset Bones between actions

![image](https://github.com/KhronosGroup/glTF-Blender-IO/assets/1696020/79aab6f3-7135-4981-9baa-1603b7b825cb)

**Comment by julienduroure** [comment]
Duplicate here: https://projects.blender.org/blender/blender/issues/139368
